### ISSUE TITLE
ReindexAll -> getState() is not correct if the Indexer broke with PHP fatal error  

### ISSUE THREAD

**Issue by larsroettig** [magento/magento2#11166]
If some fatal PHP error happens in the indexing process. The status is not correct reset
to  **StateInterface::STATUS_INVALID**

### Preconditions
* Magento 2.3.0-dev

### Steps to reproduce
- Implement Indexer -> fatal error in **executeFull()**
- execute` bin/magento  indexer:reindex **your_index_name** `

### Expected result
- Status is **not** working in **Magento\Indexer\Model\Indexer** line  404

### Actual result
- Status **is working** in **Magento\Indexer\Model\Indexer** line  404

**Comment by magento-engcom-team** [comment]
Hi @larsroettig 
Please try to describe the issue in-depth. Current description is ambiguous.
Thank you!

**Comment by larsroettig** [comment]
Hi, @magento-engcom-team, I update the descriptions. 

I recommend  changing **reindexAll** to :

```php
    /**
     * Regenerate full index
     *
     * @return void
     * @throws \Exception
     */
    public function reindexAll()
    {
        if ($this->getState()->getStatus() != StateInterface::STATUS_WORKING) {
            $state = $this->getState();
            $state->setStatus(StateInterface::STATUS_WORKING);
            $state->save();
            if ($this->getView()->isEnabled()) {
                $this->getView()->suspend();
            }
            try {
                $this->getActionInstance()->executeFull();
                $state->setStatus(StateInterface::STATUS_VALID);
                $state->save();
                $this->getView()->resume();
            } catch (\Exception $exception) {
                $state->setStatus(StateInterface::STATUS_INVALID);
                $state->save();
                $this->getView()->resume();
                throw $exception;
            } catch (\Error $error) {
                $state->setStatus(StateInterface::STATUS_INVALID);
                $state->save();
                $this->getView()->resume();
                throw $error;
            }
        }
    }
```

**Comment by magento-engcom-team** [comment]
Hi @larsroettig. Thank you for your report. 
The issue has been fixed in magento-engcom/magento2ce#1283 by @magento-engcom-team in 2.3-develop branch
Related commit(s):
 - 074285735cd6d33fd530302390ab8101e73d7e34
 - fde512eacfce36508bcec2b1c08ef381d0af3f92
 - e7d90de918be70731e2021dee757d1e7af374ab8
 - 6d4f36db79a9bee553b4acfe8c0820b5c046e6af
 - 2e83a36fbf3201bdaf6612dcfce08125687f0e28
 - 23b4aa11f5a4a0b6bfe76cd1e812d36d88ed4d83
 - b50bfa811876af007426b734019fd149def63bba
 - 04cd60da14a87727768165942c23fc1403016e41
 - 75d5a0522e8a18c2f403ea284d62e445e3d6ddc2
 - 5e02b15fcbe310cd13e385d0ae7a3e6017b56497
 - aad5ba6236ca8557072f62db524b23d68eae5de3
 - 3e321401374a7f2292bd3b2c4540db2c0af34216
 - 7898946b7ec40999b2007ee377d6b6312a3778b3
 - a173e1cd160b1d7cd24a855420317fdd7bfd9624
 - 41d0075104cb628ec1806130b6449929e8585997
 - dda19b2f6eff23e0904673fe7719c3fd935aed02
 - cba48cdfedc500b652deee50e3a3dc95e366eca3
 - 0c62caaf80c9659530d0c3dd0741108d1b0dadb0
 - 736f783b1b2fd361fb8328fd7c5cf9e7c3c172e2
 - 7569b7002878019a115aa2ea587d476a282fcd17

The fix will be available with the upcoming patch release.
